### PR TITLE
Bridged node actions to LSP commands. Enabled DB explorer view for LSP clients.

### PR DESCRIPTION
**Pull request by sdedic**
Example of how the DB explorer can work in a LSP client:
connection-delete

There are three parts in this PR:
### Bugfixes
bugfixes to the treeview feature introduced recently (Project explorer in vscode). Not important or API changes, just fixes.

### Bridge for node Actions to LSP commands
Actions in the IDE can be exported as a LSP command. The action needs to be declared in a layer.xml  in a form of `CodeActionsProvider`. Declarative support is done for easy registrations, so the layer only lists actions in file attributes (`action:<action-id>`).
The corresponding command is named `nbls:Category:ActionID` and it's expected that its argument is a node. The implementation will attempt to invoke the action on the specified node.

### DB explorer in vscode
With a simple API change that exports the Node from the `db` module the LSP server can offer `database.connections` treeview to LSP clients. Initialy `Connect`, `Disconnect` and `Delete` actions are available for connection nodes. The nodes beneath connections are for view / inspection only at the moment.

### CODE REVIEW

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/AbstractGlobalActionContext.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    public static <T> T withActionContext(Lookup context, Callable<T> callable) {
+        Lookup ctx = new ProxyLookup(Lookups.singleton(new ContextHolder(context)), Lookup.getDefault());
+        Throwable[] err = new Throwable[1];
+        Object[] result = new Object[1];
+        
+        try {
+            Lookups.executeWith(ctx, () -> {
+                try {
+                    // possibly cause an event to be fired in a global tracker
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/AbstractGlobalActionContext.java:77]: Nasty, but I don't have any better suggestion.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodeActionsProvider.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import org.openide.filesystems.FileObject;
+import org.openide.nodes.Node;
+import org.openide.util.ContextAwareAction;
+import org.openide.util.Lookup;
+import org.openide.util.lookup.ProxyLookup;
+
+/**
+ * Command provider for node-based commands. The provider handles commands with form:
+ * <code>"nbls:" Category ":" ActionID</code>. It attempts to execute an action 
+ * created by {@link Actions#forID(java.lang.String, java.lang.String)} applied on the
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/NodeActionsProvider.java:48]: I wanted to use `action:` prefix, but `nbls:` is unique...

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/TreeNodeRegistryImpl.java [hunk 3]
```diff
@@ -135,8 +137,17 @@ private synchronized TreeViewProvider registerManager(ExplorerManager em, String
                 throw new PropertyVetoException("Root change not allowed", e);
             }
         });
+        
+        ActionMap map = new ActionMap();
+        map.put(DefaultEditorKit.copyAction, ExplorerUtils.actionCopy(em));
+        map.put(DefaultEditorKit.cutAction, ExplorerUtils.actionCut(em));
+        map.put(DefaultEditorKit.pasteAction, ExplorerUtils.actionPaste(em));
+        map.put("delete", ExplorerUtils.actionDelete(em, false)); // NOI18N
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/TreeNodeRegistryImpl.java:145]: Wow. Is that all to get "delete" working!? Amazing.

**Review by JaroslavTulach — APPROVED** [review]
One of the important aspects of [good technology](http://wiki.apidesign.org/wiki/Good_Technology) is _time to market_. I guess we have a _good technology_ then, as the time you needed to implement all of this is really great!

File: java/java.lsp.server/vscode/package.json [hunk 4]
```diff
@@ -396,8 +400,21 @@
 			{
 				"command": "foundProjects.deleteEntry",
 				"title": "Delete"
+			},
+			{
+				"command": "nbls:Database:netbeans.db.explorer.action.Connect",
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/package.json:405]: Wow. Quite simple to register an action implemented in Java!

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/database-cookies.contextValues [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[org.netbeans.lib.=]
+org.netbeans.lib.ddl.DBConnection
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/database-cookies.contextValues:19]: Finally I know how to use this API! It wasn't clear to me from the previous (project logical view) review.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml [hunk 6]
```diff
@@ -21,10 +21,26 @@
 -->
 <!DOCTYPE filesystem PUBLIC "-//NetBeans//DTD Filesystem 1.2//EN" "http://www.netbeans.org/dtds/filesystem-1_2.dtd">
 <filesystem>
+    <folder name="Services">
+        <folder name="NBLS">
+            <file name="general-node-commands.instance">
+                <attr name="instanceCreate" methodvalue="org.netbeans.modules.java.lsp.server.explorer.NodeActionsProvider.forFile"/>
+                <attr name="action:org.openide.actions.DeleteAction" stringvalue="Edit"/>
+            </file>
+            <file name="database-connections-commands.instance">
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml:30]: Is this registration necessary? Can't all actions registered under `Actions/` folder be available?

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/database-cookies.contextValues (hunk 5 above)
Comment (sdedic) [reply java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/database-cookies.contextValues:19]: hm ! What a thought! They could, although I'd prefer to explicitly enable just those actions that were reviewed to use Swing UI wisely (= use UI abstractions where possible). Most of actions are really prepared to work from the server mode.

File: ide/db/src/org/netbeans/api/db/explorer/support/DatabaseExplorerUIs.java [hunk 7]
```diff
@@ -206,4 +209,22 @@ public int compare(Object dbconn1, Object dbconn2) {
             }
         }
     }
+    
+    public static Node connectionsNode() {
```
[the remark this answers is not in the document]
Comment (sdedic) [reply ide/db/src/org/netbeans/api/db/explorer/support/DatabaseExplorerUIs.java:213]: re. ExplorerManager - I thought about that and I am not sure: would it be shared = would all the clients cooperate on selection and explored context ? Creating an EM around a node is easy enough.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml (hunk 6 above)
Comment (sdedic) [reply java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml:30]: hm ! What a thought! They could, although I'd prefer to explicitly enable just those actions that were reviewed to use Swing UI wisely (= use UI abstractions where possible). Most of actions are really prepared to work from the server mode.

**Review by dbalek — APPROVED** [review]

**Comment by sdedic** [comment]
The `commit-validation` fails just with the notoriously failing `VerifySimpleTemplatesTest`, so merging. [NETBEANS-6266]
(https://issues.apache.org/jira/browse/NETBEANS-6266) filed for the failing test.

**Pull request merged**